### ISSUE TITLE
.NET Standard 2.0 build for Terminal.Gui

### ISSUE THREAD

**Issue by halvarsson** [gui-cs/Terminal.Gui#3445]
**Is your feature request related to a problem? Please describe.**
Since 1.10.1 Terminal.Gui is only build for .NET Standard 2.1, and this caused us to get locked on the previous version while running Terminal.Gui from a Powershell module that works in both Windows Powershell & in Powershell. We are doing this today in [psedit](https://github.com/ironmansoftware/psedit)

I have recently upgraded the project to run with 1.10.0, but unfortunately there seem to be some new issues that are resolved in later versions

**Describe the solution you'd like**
It would be nice if the build could also target .NET Standard 2.0, in addition to the .NET Standard 2.1

**Describe alternatives you've considered**
I have tried building Terminal.Gui under .NET Standard 2.0, and there seem to be no issues running this as far as I can tell, and it works fine on psedit with latest release of Terminal.Gui

Another option would of course be for us to produce two separate builds, one targetting .NET Framework to work on Windows Powershell, and another targetting .NET Core that would work in Powershell, or I could build it myself and just commit this to the repo, but I would like to avoid either of these solution if possible

**Additional context**
I tried reading from the various PRs/commits for changing to only build on 2.1, but I was unable to find the exact reasoning behind dropping the support for 2.0

**Comment by dodexahedron** [comment]
Could you do a git blame and find when the TFM changed? I'd be happy to go from there and try to determine why that TFM, specifically, was selected.

**Comment by halvarsson** [comment]
That would be great 

This seems to be the commit where it was changed, the project was upgraded from .NET6.0 to .NET7.0 and .NET Standard 2.0 to 2.1 https://github.com/gui-cs/Terminal.Gui/commit/a242af54e34470f1c78a503699ac5f6fc64aa475

**Comment by dodexahedron** [comment]
It's always frustrated me that netstandard 2.1 is 2.1 and not a full version number higher, _especially_ since it was for Mono, not .net Framework, and 2.0 -> 2.1 just... Really does not get across how you're about to walk into a minefield if you need to support .net Framework.

I am not certain if this will fix it at your stage of the build, but it might, depending on exactly why it hates the newer versions. If you can, please try this if you can without too much difficulty: Build against a version that you currently can't use because of netstandard2.1, but add this design-time-only polyfill reference to the csproj and let me know if it does work:

```xml
        <PackageReference Include="Meziantou.Polyfill" Version="1.0.38" PrivateAssets="all" />
```

It won't add any additional runtime dependencies - just adds the missing stuff, if possible to do so. I'm actually using it for Roslyn source generators in Terminal.Gui V2, since Visual Studio is a .net Framework host, so the generators themselves have to be netstandard2.0 even if they write code for .net 8 and C# 12. 😒

If it's just barfing at it because the TFM doesn't match, and refusing to go from there, well then that's a different story and we'll have to make it happy from here if we can.

**Comment by dodexahedron** [comment]
Also, for your modules, if you ever need PSCore features, PSCore _is_ on every supported version of Windows, today. If you launch `pwsh` before doing anything else, you'll have a guarantee you're in PSCore.

And, even if you're supporting clear back to Windows 10 1607 with .net framework 4.6.2, that version of windows also has PSCore 7.2 installed by default, so you can depend on it being there.

PSCore is required for cross-platform support, as well, anyway.

**Comment by halvarsson** [comment]
I am only targeting .NET Standard 2.0, and the reasoning for this is that this is supported from both .NET Framework and .NET Core. This allows the same PowerShell module to be loaded in both Windows PowerShell (runs on top of .NET Framework, shipped ootb on Windows 11) & PowerShell (runs on .NET 8, installed separately on Windows, Linux, MacOS etc.)

I could of course just use pwsh, but the issue is that this is needs to be installed manually, while Windows PowerShell is always available on any Windows servers where I might need it

See below, I am loading the same DLL, and get identical behavior from both (even if they are on different versions of .NET)

![image](https://github.com/gui-cs/Terminal.Gui/assets/18571127/1fb84023-3144-426f-8610-473bc025dfcc)

**Comment by dodexahedron** [comment]
OK, so basically it's just resolving things improperly, then, since we do multi-target to net472.

You can explicitly control resolution for nuget packages, instead of letting msbuild/nuget try to resolve it and end up picking the wrong one, by defining the fallback TFM used for dependency resolution. Here's what that would look like for your csproj, to make the net472 build grab the net472 nuget package specifically, and not change anything about other TFMs:

```xml
<AssetTargetFallback Condition=" '$(TargetFramework)'=='net472' ">
    net472;netstandard2.0;
</AssetTargetFallback>
```

That _should_ make it pull the one you need.

NuGet's handling of a project's `TargetFrameworks` element isn't quite as straightforward as one might assume it to be. I came across a deep dive over some specifics recently. I'll see if I can find it to link here. But basically nuget doesn't always grab individual versions for every TFM in the plural tag, for every referenced package, and it's really annoying. The above workaround usually takes care of it.

I'm not sure how it behaves when using the project.deps.json format like that, though. The recommendation from MS is to migrate to PackageReference, in any case.

But the resolution override above should in theory take care of you, if using the csproj and PackageReference for nuget packages, since the dependencies aren't hard linked into the Terminal.Gui dll. While tracing it, I was at first suspicious of the System.Management 6.0 reference in Terminal.Gui.csproj, but netstandard2.0 components are available all the way down with that. But that does add to the difficulty it could have in dependency resolution, since that package _doesn't_ explicitly define net472, nor do its dependencies. But they _do_ define netstandard2.0.

And that's why I put both net472 and netstandard2.0 in the above xml - in the hopes that it will be enough to make it pick those versions, if that or anything else is what's confusing it.

But...

Another thought just occurred to me...

We started strong name signing at some point recently, which very well could theoretically create an issue for some configurations, since that _does_ require a specific fully-qualified assembly name (which includes version) for all components, all the way up the chain.

Let me know if the workaround is the magic sauce you needed. I'm trying to find a solution that won't require another release and which is viable for anyone in your specific situation, to avoid any unexpected side-effects for anyone else.

**Comment by dodexahedron** [comment]
Also, the workaround will not work if you don't have net472 declared.

This sort of thing is one of the many reasons netstandard was abandoned and Multi-Targeting was chosen as the preferred means of doing things since then. It works when it works, and _might_ give you a single binary that runs anywhere. But when it doesn't, it's a pain to handle. And the tools all handle multi-targeting, so it shouldn't affect your users to do that instead of targeting only netstandard.

**Comment by halvarsson** [comment]
Thanks for looking into this. We only have .NET Standard 2.0 declared, and there's no users directly referencing our library, but its built as a [PowerShell Standard Library Module](https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/create-standard-library-binary-module) that's distributed from [PSGallery](https://www.powershellgallery.com/packages/psedit)

I did however find [a workaround](https://learn.microsoft.com/en-us/powershell/gallery/concepts/module-psedition-support?view=powershellget-3.x#targeting-multiple-editions) to targeting multiple editions, and it would just require us including net472 / netstandard2.1 dll in the package, but at least it would allow us to keep it as the same module (and not require multiple versions)

I mainly created this issue cause I couldn't see why .NET Standard 2.0 support was dropped, but based on your descriptions it seems you had your reasons for doing it

I tried implementing the described workaround, with .NET Standard 2.1 + .NET Framework 4.7.2 on latest version of Terminal.Gui

    <TargetFrameworks>netstandard2.1;net472</TargetFrameworks>

So I think this would be an acceptable workaround, I just need to reorganize our builds a bit to copy around the files and implement the logic for which dlls to load depending on the PowerShell edition

![image](https://github.com/gui-cs/Terminal.Gui/assets/18571127/0054bccb-55b9-4e5a-9a17-b623e7ee8698)

**Comment by dodexahedron** [comment]
Hmmm.


Well, just to potentially save you some effort:

That shouldn't, in theory, be necessary at all. Multi-targeting "just works" without manual intervention, and a consuming system will grab the version appropriate for itself.

As long as you publish your nuget and psgallery artifacts for all targets, it should be all good.

The only "drawback" to that is that, if a user uses both Windows PowerShell and PowerShell, they'll end up with two copies of the module on their machine, but that should be transparent to them anyway, and is already how most modules work that have multiple target environments.

For example, if I were to open up windows PS and then run `install-module` to install your module, it'll go out and automatically get the one defined for Windows PowerShell.

Then, if I were to open pwsh and run install-module, it'll go out and get the one for Core.

**Comment by dodexahedron** [comment]
By the way, I like the idea behind your module. Seems like a nice enhancement on top of a powershell environment. :)

**Comment by dodexahedron** [comment]
One future warning, though:

Terminal.Gui 2.0 targets .net 8 and cannot be built against earlier versions of .net because of some source and binary capabilities required that can't be polyfilled.

**Comment by halvarsson** [comment]
I think bundling multiple dlls and then conditionally load them when importing the manifest is the only way to have a multi edition module in PSGallery, its the suggested way from [documentation for PSGallery](https://learn.microsoft.com/en-us/powershell/gallery/concepts/module-psedition-support?view=powershellget-3.x#targeting-multiple-editions), and its also how Microsoft does it on f.ex. [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/Engine/Engine.csproj)

I have a PR https://github.com/ironmansoftware/psedit/pull/63 where I implemented it, but have some issues with CI that I need to figure out first. But on this branch I got it working on PowerShell 5.1 and in PowerShell 7.4 with Terminal.Gui 1.16.0

I am trying to get the module improved as much as possible before publishing a new version on PSGallery, as I recently merged in support for more languages, so it supports PowerShell, JSON and normal text so far. I have YAML support almost done and hope to add support for a few more languages. 

But we had a lot of bugs related to Terminal.Gui, which you of course fixed long ago 😊 but due to us being on the earlier versions we've been unable to resolve them. I bumped us from 1.9.0 to 1.10.0 the other day, and this fixed a couple of bugs users had reported, but then it introduced new ones 😒

From what I can see from 1.16.0 then everything we encountered is resolved, so trying to get us onto this one

In regards to Terminal.Gui 2.0, I think we would move to it at some point in the future, but currently I get most of my value from having it work in Windows PowerShell, so that's where I put my effort currently. But I assume all the work being done should translate to v2 quite easily, since the editor logic etc. is decoupled from the Terminal.Gui specific code

**Comment by dodexahedron** [comment]
> I think bundling multiple dlls and then conditionally load them when importing the manifest is the only way to have a multi edition module in PSGallery, its the suggested way from [documentation for PSGallery](https://learn.microsoft.com/en-us/powershell/gallery/concepts/module-psedition-support?view=powershellget-3.x#targeting-multiple-editions), and its also how Microsoft does it on f.ex. [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer/blob/master/Engine/Engine.csproj)
> 
> I have a PR [ironmansoftware/psedit#63](https://github.com/ironmansoftware/psedit/pull/63) where I implemented it, but have some issues with CI that I need to figure out first. But on this branch I got it working on PowerShell 5.1 and in PowerShell 7.4 with Terminal.Gui 1.16.0
> 
> I am trying to get the module improved as much as possible before publishing a new version on PSGallery, as I recently merged in support for more languages, so it supports PowerShell, JSON and normal text so far. I have YAML support almost done and hope to add support for a few more languages.
> 
> But we had a lot of bugs related to Terminal.Gui, which you of course fixed long ago 😊 but due to us being on the earlier versions we've been unable to resolve them. I bumped us from 1.9.0 to 1.10.0 the other day, and this fixed a couple of bugs users had reported, but then it introduced new ones 😒
> 
> From what I can see from 1.16.0 then everything we encountered is resolved, so trying to get us onto this one
> 
> In regards to Terminal.Gui 2.0, I think we would move to it at some point in the future, but currently I get most of my value from having it work in Windows PowerShell, so that's where I put my effort currently. But I assume all the work being done should translate to v2 quite easily, since the editor logic etc. is decoupled from the Terminal.Gui specific code

OOO! Something I learned fairly recently: If you have YAML support, you have JSON support and don't need two parsers, because YAML is formally a proper superset of JSON. Nifty.

**Comment by dodexahedron** [comment]
Why put this in v2 release? 🤔

.net 8 implements netstandard2.1, but netstandard2.1 does not cover .net 8.

**Comment by tig** [comment]
https://github.com/gui-cs/Terminal.Gui/blob/2e1b1a8df47e2c695e54e0d8fd1acb2f601be8c5/Terminal.Gui/Terminal.Gui.csproj#L23

This was done in `develop`/`main` and released in V1.17.0.

**Comment by dodexahedron** [comment]
I meant the project assignment, which has now been corrected. 👌
